### PR TITLE
ddl: support specifying expr `rand()` as column default value

### LINKED ISSUE #10377: Incompatible with MySQL 8.0 about create table with default expression
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.

```
➜  ~ tidiff 'create table ttest ( a int(10) default (rand()))'
MySQL(127.0.0.1:3306)> create table ttest ( a int(10) default (rand()))
Query OK (0.007 sec)

TiDB(127.0.0.1:4000)> create table ttest ( a int(10) default (rand()))
Error 1105: line 1 column 40 near "rand()))" (total length 48)

➜  ~ tidiff 'desc ttest'
MySQL(127.0.0.1:3306)> desc ttest
+------------------------------------------------------------+
| Field | Type    | Null | Key | Default | Extra             |
+------------------------------------------------------------+
| a     | int(10) | YES  |     | rand()  | DEFAULT_GENERATED |
+------------------------------------------------------------+
1 row in set (0.003 sec)

TiDB(127.0.0.1:4000)> desc ttest
Error 1146: Table 'test.ttest' doesn't exist
```

2. What did you expect to see?

Create table successfully.

3. What did you see instead?

Error message show above.

4. What version of TiDB are you using (`tidb-server -V` or run `select tidb_version();` on TiDB)?
```
TiDB(127.0.0.1:4000)> select tidb_version()
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| tidb_version()                                                                                                                                                                                                                                                                                                                            |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| Release Version: v2.1.8-24-g7b36cdfb6-dirty
Git Commit Hash: 7b36cdfb6e8a4c279fc360ed79cf1c499b553688
Git Branch: failpoint-2.1
UTC Build Time: 2019-04-30 08:26:17
GoVersion: go version go1.12 darwin/amd64
Race Enabled: false
TiKV Min Version: 2.1.0-alpha.1-ff3dd160846b7d1aed9079c389fc188f7f5ea13e
Check Table Before Drop: false |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.002 sec)
```

### PR DESCRIPTION
**Pull request by CbcWestwolf**
### What problem does this PR solve?

Issue Number: close #10377

Problem Summary:

Now TiDB only supports `nextVal()`, `now()`/`current_timestamp()` and literal value as a column's default value.

### What is changed and how it works?

Implement `rand()` as a default value.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [x] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [x] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Comment by xiongjiwei** [comment]
please fix the test

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -1073,6 +1073,25 @@ func getDefaultValue(ctx sessionctx.Context, col *table.Column, c *ast.ColumnOpt
...
+			}
+			col.DefaultIsExpr = true
+			var sb strings.Builder
+			restoreFlags := format.RestoreStringSingleQuotes | format.RestoreKeyWordLowercase | format.RestoreNameBackQuotes |
+				format.RestoreSpacesAroundBinaryOperation
+			restoreCtx := format.NewRestoreCtx(restoreFlags, &sb)
+			if err := c.Expr.Restore(restoreCtx); err != nil {
+				return "", false, err
+			}
+			return sb.String(), false, nil
```
Comment (xiongjiwei) [inline ddl/ddl_api.go:1091]: `rand()` function returns a random floating-point value, but here you return a string, seems it is not what the user wants

File: ddl/db_integration_test.go [hunk 2]
```diff
@@ -2031,6 +2031,23 @@ func TestDefaultValueIsString(t *testing.T) {
...
 
+func TestDefaultColumnWithRand(t *testing.T) {
+	// Related issue: https://github.com/pingcap/tidb/issues/10377
+	store, clean := testkit.CreateMockStoreWithSchemaLease(t, testLease)
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t, t1, t2")
+	tk.MustExec("create table t (c int(10) default rand())")
```
Comment (xiongjiwei) [inline ddl/db_integration_test.go:2042]: I think there will be a syntax error in MySQL, the SQL in issue is `create table t(a int default (rand()));`

File: parser/parser.y [hunk 3]
```diff
@@ -7006,6 +7046,10 @@ FunctionCallKeyword:
 	{
 		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
 	}
+|	builtinNow '(' ExpressionListOpt ')'
+	{
+		$$ = &ast.FuncCallExpr{FnName: model.NewCIStr($1), Args: $3.([]ast.ExprNode)}
+	}
```
Comment (xiongjiwei) [inline parser/parser.y:7052]: why change this

File: ddl/db_integration_test.go (hunk 2 above)
Comment (CbcWestwolf) [reply ddl/db_integration_test.go:2042]: Yes, MySQL allows `create table t(a int default (rand()));` but gives an error when `create table t(a int default rand());`.

I'll repair it later.

File: ddl/ddl_api.go [hunk 4]
```diff
@@ -1043,7 +1043,7 @@ func getDefaultValue(ctx sessionctx.Context, col *table.Column, c *ast.ColumnOpt
 	if tp == mysql.TypeTimestamp || tp == mysql.TypeDatetime {
 		switch x := c.Expr.(type) {
 		case *ast.FuncCallExpr:
-			if x.FnName.L == ast.CurrentTimestamp {
+			if x.FnName.L == ast.CurrentTimestamp || x.FnName.L == ast.Now {
```
[the remark this answers is not in the document]
Comment (CbcWestwolf) [reply ddl/ddl_api.go:1046]: It is a patch for the modification in `parser.y`.

File: ddl/ddl_api.go (hunk 1 above)
Comment (CbcWestwolf) [reply ddl/ddl_api.go:1091]: It is a mark for later inserting values into rows, which would be overwritten by calculating the expression `rand()`  in `func EvalColDefaultExpr`.

File: parser/parser.y [hunk 5]
```diff
@@ -3332,10 +3333,50 @@ ReferOpt:
  *      https://github.com/mysql/mysql-server/blob/5.7/sql/sql_yacc.yy#L6832
  */
 DefaultValueExpr:
-	NowSymOptionFraction
+	BuiltinFunction
```
[the remark this answers is not in the document]
Comment (CbcWestwolf) [reply parser/parser.y:3336]: Yes, MySQL 8.0 supports using more functions as column's default values. Next, besides `now()`, `nextval()` and `rand()`, we will support `user()`, `uuid()`,..., etc.

File: parser/parser.y [hunk 6]
```diff
@@ -3335,6 +3336,35 @@ DefaultValueExpr:
 	NowSymOptionFraction
 |	SignedLiteral
 |	NextValueForSequence
+|	BuiltinFunction
+
+BuiltinFunction:
+	identifier '(' ')'
```
Comment (xiongjiwei) [inline parser/parser.y:3342]: `create table t(a char(10) default fff())` will report syntax error in mysql.

File: ddl/db_integration_test.go [hunk 7]
```diff
@@ -2031,6 +2031,23 @@ func TestDefaultValueIsString(t *testing.T) {
...
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t, t1, t2")
+	tk.MustExec("create table t (c int(10) default (rand()))")
+	tk.MustExec("create table t1 (c int default (rand()), c1 double default (rand()))")
+	tk.MustExec("create table t2 (c int default (rand(1)), c1 double default (rand(1)))")
+	tk.MustExec("alter table t add column c1 int(10) default (rand(2))")
+	tk.MustExec("insert into t(c) values (1),(2),(3)")
+	tk.MustExec("insert into t1(c) values (1),(2),(3)")
+	tk.MustExec("insert into t2(c) values (1),(2),(3)")
```
Comment (xiongjiwei) [inline ddl/db_integration_test.go:2048]: do we have a way to test the inserted value

**Review by xiongjiwei — COMMENTED** [review]
please also add some tests in parser

File: ddl/db_integration_test.go (hunk 7 above)
Comment (CbcWestwolf) [reply ddl/db_integration_test.go:2048]: Yes, I've add the checks for the inserted values. PTAL

**Review by xiongjiwei — COMMENTED** [review]
the PR title says support the default expression `rand()`, but actually this PR adds some other functions in the parser. They are not consistent, and you may need to change one of them. If you change the title, it is better to add some tests for the other functions.

**Comment by CbcWestwolf** [comment]
> the PR title says support the default expression `rand()`, but actually this PR adds some other functions in the parser. They are not consistent, and you may need to change one of them. If you change the title, it is better to add some tests for the other functions.

To close this issue, I would like to implement only the `rand` function in this PR. The implementation of some other builtin functions would be achieved in other PRs, so I prefer to remain the parser and check the function name in `func getDefaultValue` in `ddl/ddl_api.go` if some unsupported functions are used.

**Review by xiongjiwei — APPROVED** [review]

File: parser/parser.y [hunk 8]
```diff
@@ -3335,9 +3336,33 @@ DefaultValueExpr:
 	NowSymOptionFraction
 |	SignedLiteral
 |	NextValueForSequence
+|	BuiltinFunction
```
Comment (tangenta) [inline parser/parser.y:3339]: We should also update the comments above: https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html.

File: parser/parser.y [hunk 9]
```diff
@@ -3335,9 +3336,33 @@ DefaultValueExpr:
...
+		}
+	}
 
 NowSymOptionFraction:
-	NowSym
+	'(' NowSymOptionFraction ')'
+	{
+		$$ = $2.(*ast.FuncCallExpr)
+	}
+|	NowSym
```
Comment (tangenta) [inline parser/parser.y:3365]: This change is not compatible with MySQL:

```sql
create table t (a int, b timestamp default now() on update (now()));
```

Maybe we can revert this and remove `NowSymOptionFraction` from `DefaultValueExpr`, and convert the `"NOW"` string to `"CURRENT_TIMESTAMP"` in `setFuncCallDefaultValue()`.

File: ddl/db_integration_test.go [hunk 10]
```diff
@@ -2031,6 +2031,41 @@ func TestDefaultValueIsString(t *testing.T) {
...
+	defer clean()
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t, t1, t2")
+	tk.MustExec("create table t (c int(10) default (rand()))")
+	tk.MustExec("create table t1 (c int default (rand()), c1 double default (rand()))")
+	tk.MustExec("create table t2 (c int default (rand(1)), c1 double default (rand(1)))")
+	tk.MustExec("alter table t add column c1 double default (rand(2))")
+	tk.MustExec("insert into t(c) values (1),(2),(3)")
```
Comment (tangenta) [inline ddl/db_integration_test.go:2046]: Please also add a test for add column like 
```sql
alter table t add column d int default (rand());
```

File: ddl/ddl_api.go [hunk 11]
```diff
@@ -1035,27 +1035,64 @@ func columnDefToCol(ctx sessionctx.Context, offset int, colDef *ast.ColumnDef, o
 	return col, constraints, nil
 }
 
+// getFuncCallDefaultValue
```
Comment (tangenta) [inline ddl/ddl_api.go:1038]:
```suggestion
// getFuncCallDefaultValue gets the default column value of function-call expression.
```

**Comment by tangenta** [comment]
Rest LGTM

**Pull request merged**